README
NAME
    Google::Code::Upload - upload files to a Google Code project

VERSION
    version 0.07

SYNOPSIS
        use Google::Code::Upload;
        my $gc = Google::Code::Upload->new(
            project  => 'myproject',
            username => 'mike',
            password => 'abc123',
        );
        $gc->upload(
            file        => 'README',
            summary     => 'README for myproject',
            labels      => ['Featured'],
            description => 'Hello world',
        );

DESCRIPTION
    This module allows you to programmatically upload files to Google Code
    <https://googlecode.com>.

    It ships with an executable script for use on the command line:
    googlecode_upload.pl.

METHODS
  new
    Constructs a new "Google::Code::Upload" object. Takes the following
    key-value pairs:

    username
    password (your Google Code password from
    <https://code.google.com/hosting/settings>)
    project
    ua - something that works like a LWP::UserAgent (*optional*)

  upload
    Upload the given file to Google Code. Requires the following key-value
    pairs:

    file - the filename of the file to upload
    summary - the one-line summary to give to the file (defaults to the
    filename)
    description - text describing the upload in more detail (for example,
    the changelog entry for this release)
    labels - an arrayref of labels like "Featured", "Type-Archive" or
    "OpSys-All"

    Returns the URL where the file can be downloaded if successful -
    otherwise, dies with the HTTP status line.

    You can also export the "upload" function, if you don't want to use OO
    style. Instead of key-value pairs, specify the arguments in the
    following order:

        use Google::Code::Upload qw(upload);
        upload( $file, $project_name, $username, $password, $summary, $labels, $description );

EXPORTS
    You may optionally export "upload" to use this module in a non-OO
    manner.

AVAILABILITY
    The project homepage is
    <http://search.cpan.org/dist/Google-Code-Upload/>.

    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <https://metacpan.org/module/Google::Code::Upload/>.

SOURCE
    The development version is on github at
    <http://github.com/fayland/google-code-upload> and may be cloned from
    <git://github.com/fayland/google-code-upload.git>

BUGS AND LIMITATIONS
    You can make new bug reports, and view existing ones, through the web
    interface at <https://github.com/fayland/google-code-upload/issues>.

AUTHORS
    *   Fayland Lam <fayland@gmail.com>

    *   Mike Doherty <doherty@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Fayland Lam <fayland@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Google-Code-Upload

{{$NEXT}}

0.08      2015-03-12
  - Mark as deprecated: http://google-opensource.blogspot.com/2015/03/farewell-to-google-code.html

0.07      2012-08-06
  - POD update
  - Include Term::ReadKey as a dependency

0.06      2012-06-14
  - Re-released with the right version number

0.006     2012-06-14
  - New maintainer: Mike Doherty <doherty@cpan.org>
  - Enable OO-style usage
  - Permit setting a description for the upload
  - Throw an exception when upload fails

0.05    2010-02-05
  - fix binmode issue (RT 54202)

0.04    2009-02-07
  - bug fix in googlecode_upload.pl

0.03    2009-02-07
  - cleanup the code
  - add labels prompt in googlecode_upload.pl

0.02	2009-01-12
  - ReadMode('noecho'); for password in googlecode_upload.pl

0.01    2009-01-12
  - First version, released on an unsuspecting world.


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Test::More" => "0.96"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "upload files to a Google Code project",
  "dist_author" => [
    "Fayland Lam <fayland\@gmail.com>",
    "Mike Doherty <doherty\@cpan.org>"
  ],
  "dist_name" => "Google-Code-Upload",
  "dist_version" => "0.07",
  "license" => "perl",
  "module_name" => "Google::Code::Upload",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Exporter" => 0,
    "File::Basename" => 0,
    "Getopt::Long" => 0,
    "HTTP::Request::Common" => 0,
    "LWP::Protocol::https" => 0,
    "LWP::UserAgent" => 0,
    "Pod::Usage" => 0,
    "Scalar::Util" => 0,
    "Term::ReadKey" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [
    "bin/googlecode_upload.pl"
  ]
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Google/Code/Upload.pm
package Google::Code::Upload;
use strict;
use warnings;
# ABSTRACT: upload files to a Google Code project (deprecated)
# VERSION

use File::Basename qw/basename/;
use LWP::UserAgent;
use LWP::Protocol::https;
use HTTP::Request::Common;
use Scalar::Util qw/blessed/;
use Carp;


=head1 EXPORTS

You may optionally export C<upload> to use this module in a non-OO manner.

=cut

use Exporter qw(import);
our @EXPORT_OK = qw/ upload /;

=head1 SYNOPSIS

    use Google::Code::Upload;
    my $gc = Google::Code::Upload->new(
        project  => 'myproject',
        username => 'mike',
        password => 'abc123',
    );
    $gc->upload(
        file        => 'README',
        summary     => 'README for myproject',
        labels      => ['Featured'],
        description => 'Hello world',
    );

=head1 DESCRIPTION

B<DEPRECATED>: Google code L<shut down|http://google-opensource.blogspot.com/2015/03/farewell-to-google-code.html>

This module allows you to programmatically upload files to L<Google Code|https://googlecode.com>.

It ships with an executable script for use on the command line: L<googlecode_upload.pl>.

=head1 METHODS

=head2 new

Constructs a new C<Google::Code::Upload> object. Takes the following key-value
pairs:

=over 4

=item username

=item password (your Google Code password from L<https://code.google.com/hosting/settings>)

=item project

=item ua - something that works like a L<LWP::UserAgent> (I<optional>)

=back

=cut

sub new {
    my $class = shift;
    my %args  = @_;
    $args{$_} || croak "You must provide $_" for qw(username password project);

    if ( $args{username} =~ /^(.*?)\@gmail\.com$/ ) {
        $args{username} = $1;
    }
    my $agent_string = "$class/" . (defined $class->VERSION ? $class->VERSION : 'dev');

    my $self  = {
        ua          => $args{ua} || LWP::UserAgent->new( agent => $agent_string ),
        upload_uri  => "https://$args{project}.googlecode.com/files",
        password    => $args{password},
        username    => $args{username},
    };
    return bless $self, $class;
}

=head2 upload

Upload the given file to Google Code. Requires the following key-value pairs:

=over 4

=item file - the filename of the file to upload

=item summary - the one-line summary to give to the file (defaults to the filename)

=item description - text describing the upload in more detail (for example, the
changelog entry for this release)

=item labels - an arrayref of labels like C<Featured>, C<Type-Archive> or C<OpSys-All>

=back

Returns the URL where the file can be downloaded if successful - otherwise, dies
with the HTTP status line.

You can also export the C<upload> function, if you don't want to use OO style.
Instead of key-value pairs, specify the arguments in the following order:

    use Google::Code::Upload qw(upload);
    upload( $file, $project_name, $username, $password, $summary, $labels, $description );

=cut

sub upload {
    my $self;
    my $summary;
    my $labels;
    my $file;
    my $description;

    if (blessed $_[0]) {
        $self = shift;
        my %args = @_;
        $file    = $args{file};
        $summary = $args{summary} || basename($file);
        $labels  = $args{labels} || [];
        $description = $args{description};
    }
    else {
        $file           = shift;
        my $project     = shift;
        my $username    = shift;
        my $password    = shift;
        $summary        = shift;
        $labels         = shift || [];
        $description    = shift;

        $self = __PACKAGE__->new(
            project     => $project,
            username    => $username,
            password    => $password,
        );
    }

    my $request = POST $self->{upload_uri},
        Content_Type => 'form-data',
        Content      => [
            summary  => $summary,
            ( $description ? (description => $description) : ()),
            ( map { (label => $_) } @$labels),
            filename => [$file, basename($file), Content_Type => 'application/octet-stream'],
        ];
    $request->authorization_basic($self->{username}, $self->{password});

    my $response = $self->{ua}->request($request);

    return $response->header('Location')
        if $response->code == 201;
    croak $response->status_line;
}

1;


script/googlecode_upload.pl
#!/usr/bin/perl
use strict;
use warnings;
# ABSTRACT: script to upload files to a Google Code project
# VERSION

use Getopt::Long;
use Pod::Usage;
use Term::ReadKey;
use Google::Code::Upload qw/upload/;

=head1 SYNOPSIS

    googlecode_upload.pl [options] FILE

=head1 OPTIONS

=over 4

=item B<-?>, B<--help>

=item B<s|summary>

Short description of the file

=item B<n|project>

Google Code project name

=item B<u|user>

Your Google Code Subversion username

=item B<p|pass=s>

Your Google Code Subversion password - from L<https://code.google.com/hosting/settings>

=item B<l|labels>

An optional list of labels to attach to the file

=back

=cut

my %params;
GetOptions(
    \%params,
    'help|?',
    's|summary=s',
    'n|project=s',
    'u|user=s',
    'p|pass=s',
    'l|labels=s',
);

my $file = pop @ARGV;
unless ($file) { pod2usage(1); }
-e $file or die "$file is not found\n";

unless ( exists $params{n} ) {
    print 'Please enter your project name: ';
    while ( $params{n} = ReadLine(0) ) {
        chomp($params{n});
        last if $params{n};
    }
}
unless ( exists $params{u} ) {
    print 'Please enter your googlecode.com username: ';
    while ( $params{u} = ReadLine(0) ) {
        chomp($params{u});
        last if $params{u};
    }
}
unless ( exists $params{p} ) {
    ReadMode('noecho');
    print <<'END', ' ';
** Note that this is NOT your Gmail account password! **
It is the password you use to access Google Code repositories
and can be found here: http://code.google.com/hosting/settings
your password:
END
    while ( $params{p} = ReadLine(0) ) {
        chomp($params{p});
        last if $params{p};
    }
    ReadMode 'normal';
    print "\n";
}
unless ( exists $params{s} ) {
    print 'Please enter your file summary: ';
    while ( $params{s} = ReadLine(0) ) {
        chomp($params{s});
        last if $params{s};
    }
}

my @labels;
if ( exists $params{l} ) {
    @labels = split(/\,\s*/, $params{l} );
}
else {
    print "Please enter your file labels (eg: 'Featured, Type-Source, OpSys-All'): ";
    while ( my $labels = ReadLine(0) ) {
        chomp($labels);
        @labels = split(/\,\s*/, $labels);
        last;
    }
}

my ( $status, $reason, $url ) =
    upload( $file, $params{n}, $params{u}, $params{p}, $params{s}, \@labels );

if ( $url ) {
    print "The file was uploaded successfully.\nURL: $url\n";
}
else {
    print "An error occurred. Your file was not uploaded.\nGoogle Code upload server said: $reason ($status)\n";
}


Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "upload files to a Google Code project",
  "AUTHOR" => "Fayland Lam <fayland\@gmail.com>, Mike Doherty <doherty\@cpan.org>",
  "BUILD_REQUIRES" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Test::More" => "0.96"
  },
  "CONFIGURE_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Google-Code-Upload",
  "EXE_FILES" => [
    "bin/googlecode_upload.pl"
  ],
  "LICENSE" => "perl",
  "NAME" => "Google::Code::Upload",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Exporter" => 0,
    "File::Basename" => 0,
    "Getopt::Long" => 0,
    "HTTP::Request::Common" => 0,
    "LWP::Protocol::https" => 0,
    "LWP::UserAgent" => 0,
    "Pod::Usage" => 0,
    "Scalar::Util" => 0,
    "Term::ReadKey" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.07",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



